### PR TITLE
Removing 2 timer threads from each ISOTPSoftSocket

### PR DESCRIPTION
**Pull request by epozzobon**
Removing 2 timer threads from each ISOTPSoftSocket

ISOTPSoftSocket needs timers to handle RX and TX timeouts, which were previously
implemented in a way similar to what is done in the threading.Timer object in
python: spawning a thread for each timer.

This PR removes the two timer threads from each ISOTPSoftSocket, and replaces
them with one global static thread, implemented in the TimeoutScheduler class.

This should make the ISOTPSoftSocket more performant and fix some
ISOTPSoftSocket tests that sometimes fail due to timeouts.

PS: The ISOTPScan tests are currently failing due to an unrelated issue
fixed in PR 2428. The ISOTPSoftSocket code is not used in ISOTPScan.

### CODE REVIEW

File: scapy/contrib/isotp.py [hunk 1]
```diff
@@ -813,82 +810,203 @@ def stop(self):
...
+        # Wait until the next timeout,
+        # or until event.set() gets called in another thread.
+        if to_wait > 0:
+            if TimeoutScheduler.VERBOSE:
+                print("# Thread going to sleep @ %f for %fs" %
+                      (now, to_wait))
+            interrupted = TimeoutScheduler._event.wait(to_wait)
+            new = TimeoutScheduler._time()
+            if TimeoutScheduler.VERBOSE:
+                print("# Thread awake @ %f, slept for %f, interrupted=%d" %
```
Comment (polybassa) [inline scapy/contrib/isotp.py]: You could use scapy's logging mechanism

File: scapy/contrib/isotp.py [hunk 2]
```diff
@@ -813,82 +810,203 @@ def stop(self):
...
-                        "correctly")
+    class Handle:
+        """Handle for a timeout, consisting of a callback and a time when it
+        should be executed."""
+        def __init__(self, when, cb):
+            self._when = when
+            self._cb = cb
+            self._called = False
+
+        def called(self):
```
Comment (polybassa) [inline scapy/contrib/isotp.py]: Not used so far. Maybe remove it

File: scapy/contrib/isotp.py [hunk 3]
```diff
@@ -813,82 +810,203 @@ def stop(self):
...
-        """Stop the thread, making this object unusable."""
-        if not self._dead:
-            self._killed = True
-            self._cancelled.set()
-            self._busy_sem.release()
-            self.join()
-            if not self._ready_sem.acquire(False):
-                warning("ISOTP Timer thread may not have stopped "
-                        "correctly")
+    class Handle:
```
Comment (polybassa) [inline scapy/contrib/isotp.py]: you could use `__slots__` here

File: scapy/contrib/isotp.py (hunk 1 above)
Comment (epozzobon) [reply scapy/contrib/isotp.py]: Done

File: scapy/contrib/isotp.py (hunk 2 above)
Comment (epozzobon) [reply scapy/contrib/isotp.py]: Done

File: scapy/contrib/isotp.py (hunk 3 above)
Comment (epozzobon) [reply scapy/contrib/isotp.py]: Done

**Comment by polybassa** [comment]
The unit tests of this PR will probably be fixed after #2428 is merged

**Comment by polybassa** [comment]
Could you please rebase and push. #2428 is merged, which should fix your travis build.

**Comment by polybassa** [comment]
@gpotter2 Could you please retrigger the AppVeyor build with your admin rights? It seems that this build got stuck somehow.

File: test/contrib/isotp.uts [hunk 4]
```diff
@@ -865,10 +865,8 @@ assert(succ.is_set())
 with ISOTPSocket(MockCANSocket(), sid=0x641, did=0x241) as s:
     assert(s.impl.rx_thread.is_alive())
     s.close()
-    s.impl.rx_thread.join(5)
+    assert(not s.impl.rx_thread.join())
```
Comment (gpotter2) [inline test/contrib/isotp.uts]: Could you leave a timeout in here ?

Comment (epozzobon) [reply test/contrib/isotp.uts]: I didn't remember making this change, maybe it was a failure in the merge.

**Comment by polybassa** [comment]
@gpotter2 Do you know, why AppVeyor is failing like this?

**Comment by gpotter2** [comment]
I mean, it really looks like to be freezing on a gmlan test :/ maybe a few tests before or after

**Comment by polybassa** [comment]
Thanks. I'll check if I can find something. It looked just strange to me, because all other tests passed.

**Comment by polybassa** [comment]
@gpotter2 Could you please restart the first appveyor build

**Comment by polybassa** [comment]
ready for merge, from my side

**Comment by polybassa** [comment]
@gpotter2 can this PR get merged? It would improve the stability of ISOTP tests.

**Comment by gpotter2** [comment]
I should have asked you to rebase...
this PR breaks the docs i just managed to fix :(

**Pull request merged**